contracts/router/UniversalSwap.sol

```diff
@@ -0,0 +1,324 @@
+/// The internal Token Tree allows to store up to 256 tokens, which should be enough for most use cases.
+/// Note: unlike traditional Saddle pools, tokens in UniversalSwap could be duplicated.
+/// This contract is supposed to be used in conjunction with Synapse:Bridge:
+/// - The bridged token has index == 0, and could not be duplicated in the tree.
+/// - Other tokens (correlated to bridge token) could be duplicated in the tree. Every node token in the tree
+/// is represented by a trade path from root token to node token.
+/// > This is the reason why token could be duplicated. `nUSD -> USDC` and `nUSD -> USDT -> USDC` both represent
+/// > USDC token, but via different paths from nUSD, the bridge token.
+/// In addition to the standard ISaddle interface, UniversalSwap also implements getters to observe the internal
+/// tree, as well as the best path finder between any two tokens in the tree.
+contract UniversalSwap is TokenTree, Ownable, IUniversalSwap {
+    using SafeERC20 for IERC20;
+
+    // solhint-disable-next-line no-empty-blocks
+    constructor(address bridgeToken) TokenTree(bridgeToken) {}
+
+    // ═════════════════════════════════════════════════ EXTERNAL ══════════════════════════════════════════════════════
+
+    /// @notice Adds a pool with `N = tokensAmount` tokens to the tree by adding N-1 new nodes
+    /// as the children of the given node. Given node needs to represent a token from the pool.
+    /// @dev `poolModule` should be set to address(this) if the pool conforms to ISaddle interface.
+    /// Otherwise, it should be set to the address of the contract that implements the logic for pool handling.
+    /// @param nodeIndex        The index of the node to which the pool will be added
+    /// @param pool             The address of the pool
+    /// @param poolModule       The address of the pool module
+    /// @param tokensAmount     The amount of tokens in the pool
+    function addPool(
+        uint256 nodeIndex,
+        address pool,
+        address poolModule,
+        uint256 tokensAmount
+    ) external onlyOwner {
+        require(pool != address(0), "Pool address can't be zero");
+        _addPool(nodeIndex, pool, poolModule, tokensAmount);
+    }
+
+    /// @inheritdoc IUniversalSwap
+    function swap(
+        uint8 nodeIndexFrom,
+        uint8 nodeIndexTo,
+        uint256 dx,
+        uint256 minDy,
+        uint256 deadline
+    ) external returns (uint256 amountOut) {
+        uint256 totalTokens = _nodes.length;
+        // solhint-disable-next-line not-rely-on-time
+        require(block.timestamp <= deadline, "Deadline not met");
+        require(
+            nodeIndexFrom < totalTokens && nodeIndexTo < totalTokens && nodeIndexFrom != nodeIndexTo,
+            "Swap not supported"
+        );
+        // Pull initial token from the user
+        IERC20(_nodes[nodeIndexFrom].token).safeTransferFrom(msg.sender, address(this), dx);
```

Comment:
Update to `dx` needed in case of non-standard ERC20 transfer function? As with [private pools](https://github.com/synapsecns/synapse-contracts/blob/mikeyrf/private-pool/contracts/concentrated/PrivatePool.sol#L257)

Reply:
Addressed in [c821a9e](https://github.com/synapsecns/synapse-contracts/pull/229/commits/c821a9e9fd82d85157c95099b181394c707e654a). Will add coverage for this a bit later.